src/mesh/coordinates.cxx

```diff
@@ -249,57 +268,165 @@ Coordinates::Coordinates(Mesh *mesh, const CELL_LOC loc, const Coordinates* coor
       IntShiftTorsion(mesh), localmesh(mesh), location(loc) {
 
-  dx = interpolateAndNeumann(coords_in->dx, location);
-  dy = interpolateAndNeumann(coords_in->dy, location);
+  std::string suffix = "";
+  switch (location) {
+  case CELL_XLOW: {
+      suffix = "_xlow";
+      break;
+    }
+  case CELL_YLOW: {
+      suffix = "_ylow";
+      break;
+    }
+  case CELL_ZLOW: {
+      // geometrical quantities are Field2D, so CELL_ZLOW version is the same
+      // as CELL_CENTRE
+      suffix = "";
+      break;
+    }
+  default: {
+      throw BoutException("Incorrect location passed to "
+          "Coordinates(Mesh*,const CELL_LOC,const Coordinates*) constructor.");
+    }
+  }
 
   nz = mesh->LocalNz;
 
   dz = coords_in->dz;
 
-  // Diagonal components of metric tensor g^{ij}
-  g11 = interpolateAndNeumann(coords_in->g11, location);
-  g22 = interpolateAndNeumann(coords_in->g22, location);
-  g33 = interpolateAndNeumann(coords_in->g33, location);
+  if (!force_interpolate_from_centre && mesh->sourceHasVar("dx"+suffix)) {
+
+    checkStaggeredGet(mesh, "dx", suffix);
+    if (mesh->get(dx, "dx"+suffix)) {
+      output_warn.write(
+          "\tWARNING: differencing quantity 'dx%s' not found. Set to 1.0\n", suffix.c_str());
+      dx = 1.0;
+    }
+    dx.setLocation(location);
 
-  // Off-diagonal elements.
-  g12 = interpolateAndNeumann(coords_in->g12, location);
-  g13 = interpolateAndNeumann(coords_in->g13, location);
-  g23 = interpolateAndNeumann(coords_in->g23, location);
+    if (mesh->periodicX) {
+      mesh->communicate(dx);
+    }
+
+    checkStaggeredGet(mesh, "dy", suffix);
+    if (mesh->get(dy, "dy"+suffix)) {
+      output_warn.write(
+          "\tWARNING: differencing quantity 'dy%s' not found. Set to 1.0\n", suffix.c_str());
+      dy = 1.0;
+    }
+    dy.setLocation(location);
+
+    // grid data source has staggered fields, so read instead of interpolating
+    // Diagonal components of metric tensor g^{ij} (default to 1)
+    getAtLoc(mesh, g11, "g11", suffix, location, 1.0);
+    getAtLoc(mesh, g22, "g22", suffix, location, 1.0);
+    getAtLoc(mesh, g33, "g33", suffix, location, 1.0);
+    getAtLoc(mesh, g12, "g12", suffix, location, 0.0);
+    getAtLoc(mesh, g13, "g13", suffix, location, 0.0);
+    getAtLoc(mesh, g23, "g23", suffix, location, 0.0);
+
+    /// Find covariant metric components
+    // Check if any of the components are present
+    if (mesh->sourceHasVar("g_11"+suffix) or mesh->sourceHasVar("g_22"+suffix) or
+        mesh->sourceHasVar("g_33"+suffix) or mesh->sourceHasVar("g_12"+suffix) or
+        mesh->sourceHasVar("g_13"+suffix) or mesh->sourceHasVar("g_23"+suffix)) {
+      // Check that all components are present
+      if (mesh->sourceHasVar("g_11"+suffix) and mesh->sourceHasVar("g_22"+suffix) and
+          mesh->sourceHasVar("g_33"+suffix) and mesh->sourceHasVar("g_12"+suffix) and
+          mesh->sourceHasVar("g_13"+suffix) and mesh->sourceHasVar("g_23"+suffix)) {
```

Comment:
I think these two `if` statements could be simplified with `std::any_of` and `std::all_of`:

```cpp
auto metric_components = {"g_11", "g_22", ...};
auto source_has_component = [&suffix] (const std::string& name) {
  return mesh->sourceHasVar(name + suffix);
};
if (std::any_of(begin(metric_components), end(metric_components),
                      source_has_component)) {
  if (std::all_of(begin(metric_components), end(metric_components),
                        source_has_component)) {
```

Ok, still a tiny bit verbose, but the logic is maybe more explicit/readable